man/man1/pydbase.1
.TH pydbase 1 "" "" \" -*- nroff -*-
.SH NAME
.B pydbase \-
Simple python database
.SH SYNOPSIS
.B pydbase [options] [key data]
.br
.SH DESCRIPTION
This manual page documents the pydbase python database
.PP
pydbase Features:
.br
    o Simple usage, reliable storage
.br
    o Per record checksum
.br
    o Multi platform operation (Linux and Windows for now)
.br
    o Fast. Because it is simple
.br
.SH COMMAND LINE OPTIONS

.br
Usage: dbaseadm.py [options] [arg_key arg_data]
.br
    -h          Help (this screen)
.br
    -E          Replace record in place (see notes)
.br
    -V          Print version
.br
    -q          Quiet on, less printing
.br
    -d          Debug level (0-10)
.br
    -v          Increment verbosity level
.br
    -r          Randomize (with -w)
.br
    -w          Write random record(s)
.br
    -z          Dump backwards(s)
.br
    -i          Show deleted record(s)
.br
    -U          Vacuum DB
.br
    -R          Re-index / recover DB
.br
    -I          DB Integrity check
.br
    -c          Set check integrity flag
.br
    -s          Skip to count recs
.br
    -K          List keys only
.br
    -S          Print num recs
.br
    -m          Dump data to console
.br
    -n  num     Num of recs (with -w)
.br
    -t  keyval  Retrieve by key
.br
    -o  offs    Get data from offset
.br
    -G  num     Get record by number
.br
    -F  subkey  Find by sub str
.br
    -g  num     Get number of recs.
.br
    -k  keyval  Key to save
.br
    -a  str     Data to save
.br
    -y  keyval  Find by key
.br
    -D  keyval  Delete by key
.br
    -p  num     Skip number of recs
.br
    -u  recnum  Delete at recnum
.br
    -l  lim     Limit get records
.br
    -e  offs    Delete at offset
.br
    -Z  keyval  Get record position
.br
    -X  max     Limit recs on delete
.br
    -f  file    DB file for save/retrieve default: 'pydbase.pydb'

The verbosity / debug  level influences the amount of printout presented.
Use quotes for multi word arguments.

Notes:

 For in-place replacement, the replacement record has to be shorter than \
 the existing record. If longer, the record is stored as normal.

.SH BLOCKCHAIN

 The database is extended with a blockhcain implementation. The new class
is called twinchain; and it is a class derived from twincore.

 To drive the blockchain, use chainad.py append method. The database will calculate
all the hashes, integrate it into the existing chain with the new item getting
a back link field. This field is calculated based upon the previous record's
hash and the previous record's frozen date. This assures that identical data
will have a different hash, so data cannot be anticipated based upon its hash
alone. The hash is done with 256 bits, and assumed to be very secure.

Usage: chainadm.py [options]
.br
    -a  data   append data to the end of chain
.br
    -g recnum  get record
.br
    -k reckey  get record by key/header
.br
    -G recnum  get record offset
.br
    -r recnum  get record header
.br
    -n         append / show / get number of records
.br
    -e         override header (checked for UUID)
.br
    -t         print record's UUID date)
.br
    -s         skip count
.br
    -x         max record count to list
.br
    -m         dump chain data
.br
    -c         check data integrity
.br
    -i         check link integrity
.br
    -S         get db size
.br
    -v         increase verbosity
.br
    -h         help (this screen)

.SH FILESPEC

Saved file specification consists of two parts:
.PP
    o Data file
.br
    o Index file
.PP
Structure of the data:

    DATA FILE HEADER
    FILESIG     = b"PYDB"
    32 byte header, starting with FILESIG

    4 bytes    4 bytes          4 bytes         Variable
    ------------------------------------------------------------
    RECSIG     Hash_of_key      Len_of_key      DATA_for_key
    RECSEP     Hash_of_payload  Len_of_payload  DATA_for_payload
        .
        .
    RECSIG     Hash_of_key      Len_of_key      DATA_for_key
    RECSEP     Hash_of_payload  Len_of_payload  DATA_for_payload

    where:

    RECSIG="RECB" (record begin here)
    RECSEP="RECS" (record separated here)
    RECDEL="RECX" (record deleted)

    Deleted records are marked with the RECSIG mutated from RECB to RECX

    Vacuum will remove the deleted records; Make sure your database has no
    pending ops; or non atomic opts when vacuuming;

Structure of the index:

    INDEX FILE HEADER

    IDXSIG      = b"PYIX"

    32 byte header, starting with IDXSIG

    32 bit index
    .
    .
    32 bit index ...

    The index is the offset of the record in the data file.

.SH PLATFORMS
All platforms that have python should work fine. Current platforms include
Linux, WinX, MSYS2 ...
.SH FILES
 pydbase/*, dbaseadm.py chainadm.py twincore.py, chaincore.py twinbase.py
 dbutils.py
.SH COPYING
pydbase can be copied under the MIT or General Public License
.SH AUTHOR
pydbase was written by Peter Glen, peterglen99@google.com



pydbase.py
#!/usr/bin/env python3

import  os, sys, getopt, signal, select, socket, time, struct
import  random, stat, os.path, datetime, threading, warnings
import  string

import pyvpacker

import gettext
gettext.bindtextdomain('thisapp', './locale/')
gettext.textdomain('thisapp')
_ = gettext.gettext

base = os.path.dirname(os.path.realpath(__file__))
sys.path.append('pydbase')

import twincore

# ------------------------------------------------------------------------

gl_lockname = "pydbase.main.lock"

# Module variables (pushed to a class)

class _m():
    pgdebug = 0
    verbose = 0
    keyonly = 0
    ncount  = 1
    skipcnt = 0
    maxx    = 10
    lcount  = twincore.INT_MAX
    quiet   = 0; writex  = 0
    randx   = 0; skipx   = 0
    offsx   = 0; delx    = 0
    delrx   = 0; delrx2  = 0
    backx   = 0; sdelx   = 0
    vacx    = 0; recx    = 0
    integx  = 0; checkx  = 0
    sizex   = 0; findx   = ""
    retrx   = ""; getit  = ""
    keyx    = ""; datax  = ""
    dkeyx   = ""; dumpx  = 0
    findrec = ""; getrec = 0

    deffile = "pydbase.pydb"

version = "0.9.3"
vdate   =  "Sat 11.Feb.2023"
allstr  =    " " + \
                string.ascii_lowercase +  string.ascii_uppercase +  \
                    string.digits

# ------------------------------------------------------------------------

# Return a random string based upon length

def randstr(lenx):

    strx = ""
    for aa in range(lenx):
        ridx = random.randint(0, len(allstr)-1)
        rr = allstr[ridx]
        strx += str(rr)
    return strx

def help():
    print("Usage: pydebase.py [options] [arg_key arg_data]")
    print(" Options: -h         help (this screen)   -|-  -i  show deleted on dump")
    print("          -V         print version        -|-  -q  quiet on")
    print("          -d         debug level (0-10)   -|-  -v  increment verbosity level")
    print("          -r         randomize data       -|-  -w  write fixed record(s)")
    print("          -z         dump backwards(s)    -|-  -i  show deleted record(s)")
    print("          -U         Vacuum DB            -|-  -R  reindex / recover DB")
    print("          -I         DB Integrity check   -|-  -c  set check integrity flag")
    print("          -s         Skip to count recs   -|-  -K  list keys only")
    print("          -y  key    find by key          -|-  -m  dump data to console")
    print("          -o  offs   get data from offset -|-  -e  offs   delete at offset")
    print("          -u  rec    delete at position   -|-  -g  num    get number of recs.")
    print("          -k  key    key to save          -|-  -a  str    data to save ")
    print("          -S         print num recs       -|-  -D  key    delete by key ")
    print("          -n  num    number of records    -|-  -t  key    retrieve by key")
    print("          -p  num    skip number of recs  -|-  -F  subkey find by sub str")
    print("          -l  lim    limit number of recs -|-  -G  num  get record number ")
    print("          -x  max    limit max number of records to get")
    print("          -f  file   input or output file (default: 'pydbase.pydb')")
    print("The verbosity level influences the amount of data presented.")
    print("On the command line, use quotes for multi word arguments.")

def mainfunc():

    ''' Exersize all funtions of the twincore library '''

    opts = []; args = []

    # Old fashioned parsing
    opts_args   = "a:d:e:f:g:k:l:n:o:s:t:u:x:y:p:D:F:G:"
    opts_normal = "mchiVrwzvqURIK?S"
    try:
        opts, args = getopt.getopt(sys.argv[1:],  opts_normal + opts_args)
    except getopt.GetoptError as err:
        print(_("Invalid option(s) on command line:"), err)
        sys.exit(1)

    # Scan twice so verbose shows up early
    for aa in opts:
        if aa[0] == "-h" or aa[0] == "-?":
            help(); exit(1)
        if aa[0] == "-v":
            _m.verbose += 1
        if aa[0] == "-d":
            try:
                _m.pgdebug = int(aa[1])
                #print( sys.argv[0], _("Running at debug level"),  pgdebug)
            except:
                _m.pgdebug = 0

    for aa in opts:
        if aa[0] == "-V":
            print("Script Version:", version);
            print("Engine Version:", twincore.version);

            if _m.verbose > 0:
                print("Compiled:", vdate);
            exit(1)

        # Action flags, one at a time
        if aa[0] == "-z":
            _m.backx = True
        if aa[0] == "-u":
            _m.delrx2 = 1
            _m.delrx = int(aa[1])
        if aa[0] == "-w":
            _m.writex = True
        if aa[0] == "-i":
            _m.sdelx = True
        if aa[0] == "-q":
            _m.quiet = True
        if aa[0] == "-n":
            _m.ncount = int(aa[1])
        if aa[0] == "-t":
            _m.retrx = aa[1]
        if aa[0] == "-l":
            _m.lcount = int(aa[1])
        if aa[0] == "-x":
            _m.maxx = int(aa[1])
        if aa[0] == "-s":
            _m.skipcnt = int(aa[1])
        if aa[0] == "-f":
            _m.deffile = aa[1]
        if aa[0] == "-g":
            _m.getit = aa[1]
        if aa[0] == "-G":
            _m.getrec = aa[1]
        if aa[0] == "-k":
            _m.keyx = aa[1]
        if aa[0] == "-D":
            _m.dkeyx = aa[1]
        if aa[0] == "-a":
            _m.datax = aa[1]
        if aa[0] == "-r":
            _m.randx = True
        if aa[0] == "-U":
            _m.vacx = True
        if aa[0] == "-K":
            _m.keyonly = True
        if aa[0] == "-R":
            _m.recx = True
        if aa[0] == "-p":
            _m.skipx = int(aa[1])
        if aa[0] == "-y":
            _m.findx = aa[1]
        if aa[0] == "-o":
            _m.offsx = aa[1]
        if aa[0] == "-e":
            _m.delx = aa[1]
        if aa[0] == "-c":
            _m.checkx = True
        if aa[0] == "-S":
            _m.sizex = True
        if aa[0] == "-I":
            _m.integx = True
        if aa[0] == "-m":
            _m.dumpx = True
        if aa[0] == "-F":
            _m.findrec = aa[1]

    #print("args", len(args), args)

    if len(args) == 1:
        print("Must have zero or two arguments. Use -h for help.")
        sys.exit(1)

    # Set some flags
    twincore.base_quiet     = _m.quiet
    twincore.base_pgdebug   = _m.pgdebug
    twincore.base_showdel   = _m.sdelx
    twincore.base_integrity = _m.checkx
    twincore.base_pgdebug   = _m.pgdebug

    # Create our database
    core = twincore.TwinCore(_m.deffile, _m.pgdebug)
    core.base_verbose   = _m.verbose

    # See if we have arguments, save it as data
    if len(args) == 2:
        #print("args", args)
        curr = core.save_data(args[0], args[1])
        sys.exit(0)

    #print(dir(core))

    # Correct maxx
    if _m.maxx == 0 : _m.maxx = 1

    dbsize = core.getdbsize()
    #print("DBsize", dbsize)

    if _m.keyx and _m.datax:
        curr = 0
        if _m.verbose:
            print("adding", _m.keyx, _m.datax)
        for aa in range(_m.ncount):
            curr = core.save_data(_m.keyx, _m.datax)
        #print("curr", curr)
    elif _m.keyx:
        curr = 0
        if verbose:
            print("adding", keyx)
        for aa in range(ncount):
            curr = core.save_data(keyx, "dddd dddd")
        #print("curr", curr)
    elif _m.writex:
        curr = 0;
        if _m.randx:
            for aa in range(_m.ncount):
                curr = core.save_data(
                    randstr(random.randint(2, 10)),
                        randstr(random.randint(10, 20)))
        else:
            for aa in range(_m.ncount):
                curr = core.save_data("111 222", "333 444")
        #print("curr", curr)
    elif _m.findx:
        if _m.lcount == 0: _m.lcount = 1
        ddd = core.find_key(_m.findx, _m.lcount)
        print("Found records:", ddd)
    elif _m.getrec:
        ddd = core.get_rec(int(_m.getrec))
        print("Got:", ddd)

    elif _m.keyonly:
        cnt = 0
        if _m.lcount + _m.skipx > dbsize:
            _m.lcount = dbsize - _m.skipx
        for aa in range(_m.skipx, _m.lcount):
            ddd = core.get_rec(aa)
            print(aa, ddd)
            cnt += 1

    elif _m.getit:
        getx = int(_m.getit)
        #skipx = dbsize - skipx
        if getx + _m.skipx > dbsize:
            getx = dbsize - _m.skipx
        if _m.skipx < 0:
            _m.skipx = 0
            print("Clipping to dbsize of", dbsize)
        if _m.verbose:
            print("Getting %d records" % getx);
            if _m.skipx:
                print("With skipping %d records" % skipx);

        for aa in range(_m.skipx, getx + _m.skipx):
            ddd = core.get_rec(aa)
            print(aa, ddd)

    elif _m.retrx != "":
        if _m.ncount == 0: _m.ncount = 1
        ddd = core.retrieve(_m.retrx, _m.ncount)
        if not ddd:
            print("Record:", "'" + _m.retrx + "'", "is not found.")
        else:
            print(ddd)
    elif _m.sizex:
        print("Database size:", core.getdbsize())
    elif _m.offsx:
        ddd = core.get_rec_offs(int(_m.offsx))
        print(ddd)
    elif _m.delx:
        ddd = core.del_rec_offs(int(_m.delx))
        print(ddd)
    elif _m.delrx2:
        ddd = core.del_rec(int(_m.delrx))
        print(ddd)
    elif _m.dkeyx:
        ddd = core.del_rec_bykey(_m.dkeyx)
        print("Deleted:", ddd, "records.")
    elif _m.recx:
        ddd = core.reindex()
        print("Reindexed:", ddd, "record(s)")
    elif _m.vacx:
        ddd = core.vacuum()
        print("Vacuumed:", ddd[0], "saved", ddd[1], "record(s)")
    elif _m.integx:
        ddd = core.integrity_check()
        print("Integrity check found good:", ddd[0], "of", ddd[1], "record(s)")
    elif _m.dumpx:
        if _m.backx:
            core.dump_data(_m.lcount, _m.skipx)
        else:
            core.revdump_data(_m.lcount, _m.skipx)
    elif _m.findrec:
            ret = core.findrec(_m.findrec, _m.lcount, _m.skipx)
            if _m.verbose:
                print("Found:", end = "")
            print(ret)
    else:
        print("Use: pydbase.py -h to see options and help")

if __name__ == "__main__":

    # Tested with process based lock (OK)
    #twincore.waitlock(gl_lockname)
    mainfunc()
    #twincore.dellock(gl_lockname)

# EOF
